src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -437,10 +437,25 @@ private static bool TryUnfoldITestDataSource(ITestDataSource dataSource, TestDat
         var discoveredTests = new List<UnitTestElement>();
         int index = 0;
 
-        foreach (object?[] d in data)
+        foreach (object?[] dataOrTestDataRow in data)
         {
+            object?[] d = dataOrTestDataRow;
+            string? displayNameFromTestDataRow = null;
+            if (d?.Length == 1 && d[0]?.GetType() is { IsGenericType: true } genericType &&
+                genericType.GetGenericTypeDefinition() == typeof(TestDataRow<>))
+            {
+                object testDataRow = d[0]!;
+                object? dataFromTestDataRow = genericType.GetProperty("Value")!.GetValue(testDataRow);
+                testDataSourceIgnoreMessage = genericType.GetProperty("IgnoreMessage")!.GetValue(testDataRow) as string ?? testDataSourceIgnoreMessage;
+                displayNameFromTestDataRow = genericType.GetProperty("DisplayName")!.GetValue(testDataRow) as string;
```

Comment:
I would introduce an interace `ITestDataRow` so we can simplify all of that and not be forced to do reflection. For the `Value`, I don't recall exactly how I did for TATF but I was transforming it into `object[]`.